File: packages/monty-expo/android/src/main/java/expo/modules/montyexpo/MontyExpoModule.kt
package expo.modules.montyexpo

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONArray
import org.json.JSONObject

private const val MODULE_VERSION = "0.1.0"

private fun toJsonValue(value: Any?): Any? {
  return when (value) {
    null -> JSONObject.NULL
    is Boolean, is Number, is String -> value
    is Map<*, *> -> {
      val obj = JSONObject()
      value.forEach { (k, v) ->
        if (k is String) {
          obj.put(k, toJsonValue(v))
        }
      }
      obj
    }
    is List<*> -> {
      val arr = JSONArray()
      value.forEach { arr.put(toJsonValue(it)) }
      arr
    }
    else -> value.toString()
  }
}

private fun mapToJson(map: Map<String, Any?>?): String? {
  if (map == null) {
    return null
  }
  val json = JSONObject()
  map.forEach { (k, v) ->
    json.put(k, toJsonValue(v))
  }
  return json.toString()
}

private fun jsonToAny(value: Any?): Any? {
  return when (value) {
    JSONObject.NULL -> null
    is JSONObject -> {
      val map = mutableMapOf<String, Any?>()
      val keys = value.keys()
      while (keys.hasNext()) {
        val key = keys.next()
        map[key] = jsonToAny(value.get(key))
      }
      map
    }
    is JSONArray -> {
      val list = mutableListOf<Any?>()
      for (i in 0 until value.length()) {
        list.add(jsonToAny(value.get(i)))
      }
      list
    }
    else -> value
  }
}

private fun parseRustResponse(raw: String): Map<String, Any?> {
  return try {
    @Suppress("UNCHECKED_CAST")
    jsonToAny(JSONObject(raw)) as? Map<String, Any?> ?: mapOf(
      "ok" to false,
      "error" to mapOf(
        "typeName" to "RuntimeError",
        "message" to "Rust response is not a JSON object",
        "traceback" to emptyList<Any?>()
      )
    )
  } catch (_: Throwable) {
    mapOf(
      "ok" to false,
      "error" to mapOf(
        "typeName" to "RuntimeError",
        "message" to "Rust response is not valid JSON",
        "traceback" to emptyList<Any?>()
      )
    )
  }
}

class MontyExpoModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MontyExpo")

    Function("version") {
      MODULE_VERSION
    }

    Function("isNativeRuntimeLinked") {
      true
    }

    Function("runSync") { code: String, runOptions: Map<String, Any?>?, montyOptions: Map<String, Any?>? ->
      try {
        val raw = MontyRustBridge.nativeRun(
          code,
          mapToJson(runOptions),
          mapToJson(montyOptions)
        )
        parseRustResponse(raw)
      } catch (e: Throwable) {
        mapOf(
          "ok" to false,
          "error" to mapOf(
            "typeName" to "RuntimeError",
            "message" to (e.message ?: "Rust bridge call failed"),
            "traceback" to emptyList<Any?>()
          )
        )
      }
    }
  }
}

File: packages/monty-expo/ios/MontyExpoModule.swift
import Foundation
import ExpoModulesCore

private let moduleVersion = "0.1.0"

@_silgen_name("monty_expo_run_json")
private func monty_expo_run_json(
  _ code: UnsafePointer<CChar>?,
  _ runOptionsJson: UnsafePointer<CChar>?,
  _ montyOptionsJson: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>?

@_silgen_name("monty_expo_string_free")
private func monty_expo_string_free(_ pointer: UnsafeMutablePointer<CChar>?)

private func toJSONString(_ value: Any?) -> String? {
  guard let value else {
    return nil
  }
  guard JSONSerialization.isValidJSONObject(value) else {
    return nil
  }
  guard let data = try? JSONSerialization.data(withJSONObject: value), let text = String(data: data, encoding: .utf8) else {
    return nil
  }
  return text
}

private func parseRustResponse(_ text: String) -> [String: Any] {
  guard let data = text.data(using: .utf8) else {
    return [
      "ok": false,
      "error": [
        "typeName": "RuntimeError",
        "message": "Rust response is not valid UTF-8",
        "traceback": []
      ]
    ]
  }
  guard let object = try? JSONSerialization.jsonObject(with: data), let dict = object as? [String: Any] else {
    return [
      "ok": false,
      "error": [
        "typeName": "RuntimeError",
        "message": "Rust response is not valid JSON",
        "traceback": []
      ]
    ]
  }
  return dict
}

private func withOptionalCString<T>(_ value: String?, body: (UnsafePointer<CChar>?) -> T) -> T {
  guard let value else {
    return body(nil)
  }
  return value.withCString { ptr in
    body(ptr)
  }
}

public class MontyExpoModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MontyExpo")

    Function("version") {
      return moduleVersion
    }

    Function("isNativeRuntimeLinked") {
      return true
    }

    Function("runSync") { (code: String, runOptions: [String: Any]?, montyOptions: [String: Any]?) -> [String: Any] in
      let runOptionsJson = toJSONString(runOptions)
      let montyOptionsJson = toJSONString(montyOptions)

      let resultPointer = code.withCString { codePointer in
        withOptionalCString(runOptionsJson) { runOptionsPointer in
          withOptionalCString(montyOptionsJson) { montyOptionsPointer in
            monty_expo_run_json(codePointer, runOptionsPointer, montyOptionsPointer)
          }
        }
      }

      guard let resultPointer else {
        return [
          "ok": false,
          "error": [
            "typeName": "RuntimeError",
            "message": "monty_expo_run_json returned nil",
            "traceback": []
          ]
        ]
      }

      let text = String(cString: resultPointer)
      monty_expo_string_free(resultPointer)
      return parseRustResponse(text)
    }
  }
}
